### PR TITLE
Rework setting PATH variable to not store in a File temporarily

### LINKED ISSUE #230: Multiple leading backslashes of Windows network path incorrectly removed from preference for toolchain folder
### Description

Global tools path preference for toolchain folder does not handle Windows network paths (with two leading backslashes) correctly.

My GNU ARM Eclipse tools are installed on a network location, let's call it \\MyServer\gnu\cross-tools. When adding this as toolchain folder under
Preferences -> C/C++ -> Build -> Global Tools Path -> Toolchain folder
(or setting ilg.gnuarmeclipse.managedbuild.cross/toolchain.path.1871385609)
I end up with this error message:
Error: Program "aarch64-none-elf-g++" not found in PATH
PATH=[[...some entries...];\MyServer\gnu\cross-tools;[...more entries...]]

It seems that the raw value of the preference is modified in some way before the PATH variable is extended causing the two leading backslashes to be merged into one, which results in an incorrect entry in the PATH variable.

I tried a few things to work around this issue but was unlucky so far:
\\\MyServer\gnu\cross-tools
\\\\MyServer\gnu\cross-tools
\\\\\\\\MyServer\gnu\cross-tools
//MyServer\gnu\cross-tools
\/MyServer\gnu\cross-tools
/\MyServer\gnu\cross-tools
All of the above result in "\MyServer\gnu\cross-tools" to end up in the PATH variable.

### Steps to Reproduce

1. Go to Preferences -> C/C++ -> Build -> Global Tools Path -> Toolchain folder and set Toolchain path to \\MyServer\gnu\cross-tools
2. Try to build a CDT project
3. The build fails because the toolchain is not found in the PATH

**Expected behaviour:** The plugin should not merge two leading backslashes before extending the PATH variable.

**Actual behaviour:** The plugin incorrectly merges all leading backslashes of the Toolchain path value before extending the PATH variable.

### Versions

* [plug-in version] ilg.gnuarmeclipse.managedbuild.cross 2.2.1.201606210758
* [Eclipse version] Eclipse 4.6.1
* [Java version] Java 1.8
* [operating system] Windows Server 2012 R2 Standard
* [toolchain version] gcc-linaro-aarch64-none-elf-4.9-2014.05

### PR DESCRIPTION
**Pull request by jonahgraham**
The old code placed multiple entries for adding to PATH in a File temporarily to get the absolute path. This worked generally on non-Windows platforms because often there was only a single entry.

However on Windows there are generally two entries. Adding to that is the UNC path names that are sometimes present on Windows and the temporary use of File causes corruption in the settings.

Instead collect the new PATH entries in a list of strings, and then before making the final PATH setting, run each individual entry through new File(entry).getAbsolutePath()

TODO: 

- [x] needs testing - this first version is a quick attempt to understand problem
- [x] I wrote this on Linux, needs testing on some other platforms
- [x] Needs applying to the RISC-V class of the same name

Fixes #230

### CODE REVIEW

File: plugins/org.eclipse.embedcdt.managedbuild.cross.arm.core/src/org/eclipse/embedcdt/managedbuild/cross/arm/core/EnvironmentVariableSupplier.java [hunk 1]
```diff
@@ -139,34 +133,28 @@ public static PathEnvironmentVariable create(IConfiguration configuration) {
...
 
-				File sysroot = new File(path);
-				// File bin = new File(sysroot, "bin"); //$NON-NLS-1$
-				// if (bin.isDirectory()) {
-				// sysroot = bin;
-				// }
 				if (DEBUG_PATH) {
 					if (Activator.getInstance().isDebugging()) {
-						System.out.println("arm.PathEnvironmentVariable.create() PATH=\"" + sysroot + "\" cfg="
+						System.out.println("arm.PathEnvironmentVariable.create() PATH=\"" + path + "\" cfg="
```
Comment (jonahgraham) [inline plugins/org.eclipse.embedcdt.managedbuild.cross.arm.core/src/org/eclipse/embedcdt/managedbuild/cross/arm/core/EnvironmentVariableSupplier.java:153]: Probably should be `pathVar`
```suggestion
						System.out.println("arm.PathEnvironmentVariable.create() PATH=\"" + pathVar + "\" cfg="
```

**Comment by ilg-ul** [comment]
Should we be concerned with compatibility issues? My first impulse is to leave the buggy code there, with a deprecation notice, and add a new class with the fix.

**Comment by jonahgraham** [comment]
> Should we be concerned with compatibility issues? My first impulse is to leave the buggy code there, with a deprecation notice, and add a new class with the fix.

I'll have a think about that more while I write and test it. The external behaviour of the code isn't changing with my fix, except for places where the entries are being corrupted. Therefore my current instinct is to simply fix the issue.

**Comment by ilg-ul** [comment]
If in the end only the private class or private members are changed, you are right, we should have no compatibility issues.

**Comment by ilg-ul** [comment]
Do you plan more work on this PR?

**Comment by jonahgraham** [comment]
> Do you plan more work on this PR?

Yes - I will finish this up in the coming days.

File: plugins/org.eclipse.embedcdt.managedbuild.cross.riscv.core/src/org/eclipse/embedcdt/managedbuild/cross/riscv/core/EnvironmentVariableSupplier.java [hunk 2]
```diff
@@ -89,28 +92,21 @@ public static PathEnvironmentVariable create(IConfiguration configuration) {
...
 
-			String path = "";
+			List<String> path = new ArrayList<>();
 			if (preferXpacksBin) {
-				IPath projectPath = project.getWorkspace().getRoot().getLocation().append(project.getFullPath());
-				IPath xpackBinPath = projectPath.append("xpacks").append(".bin");
-
-				path = xpackBinPath.toOSString();
+				String xpackBinPath = project.getFolder("xpacks").getFolder(".bin").getLocation().toOSString();
+				path.add(xpackBinPath);
```
Comment (ilg-ul) [inline plugins/org.eclipse.embedcdt.managedbuild.cross.riscv.core/src/org/eclipse/embedcdt/managedbuild/cross/riscv/core/EnvironmentVariableSupplier.java:98]: Does this produce the absolute path of the `xpacks/.bin` folder located inside the project?

**Comment by jonahgraham** [comment]
I have tested this on Linux and Windows. On Windows since I don't actually have a server to install tools I used `\\machine\c$`. So instead of:

`C:/Users/ditto-win/AppData/Roaming/xPacks/@xpack-dev-tools/windows-build-tools/4.4.0-1.1/.content/bin`

I used:

`\\ditto-win\c$\Users\ditto-win\AppData\Roaming\xPacks\@xpack-dev-tools\windows-build-tools\4.4.0-1.1\.content\bin`

(`ditto-win` is both the name of the Windows computer I am using and the user name)

File: plugins/org.eclipse.embedcdt.managedbuild.cross.riscv.core/src/org/eclipse/embedcdt/managedbuild/cross/riscv/core/EnvironmentVariableSupplier.java (hunk 2 above)
Comment (jonahgraham) [reply plugins/org.eclipse.embedcdt.managedbuild.cross.riscv.core/src/org/eclipse/embedcdt/managedbuild/cross/riscv/core/EnvironmentVariableSupplier.java:98]: This fixed a bug that the code assumed that the project was located on disk in the workspace folder.

e.g. if I have a project on my computer in `E:\cdt\git\armproj` and my workspace is `E:\cdt\runtime-CDT` then the old code would add `E:\cdt\runtime-CDT\armproj\xpacks\.bin` to the `path` variable, when it should have added `E:\cdt\git\armproj\xpacks\.bin`

This was happening because of the order of resolution. The updated code gets the full internal path to the `bin` first (e.g. `project.getFolder("xpacks").getFolder(".bin")` == `F/armproj/xpacks/.bin`) and then gets the location on disk of that folder. The old code (`project.getWorkspace().getRoot().getLocation()`) got the location on disk of the workspace and then appended paths to it.

Let me know if you want me to split that out to a different PR.

Comment (jonahgraham) [reply plugins/org.eclipse.embedcdt.managedbuild.cross.riscv.core/src/org/eclipse/embedcdt/managedbuild/cross/riscv/core/EnvironmentVariableSupplier.java:98]: > Does this produce the absolute path of the xpacks/.bin folder located inside the project?

TL;DR - Yes! :)

**Comment by ilg-ul** [comment]
In your example you show the path for the windows-build-tools, which is retrieved from the preferences as an absolute path.

My question refers to the `xpacks/.bin` folder. Do you see the absolute path of this folder in the generated PATH?

This is a relatively recent feature, unfortunately not very well documented. :-(

`xpm` can also install toolchains inside a project, in the `xpacks/.bin` folder, and in this case the local toolchain is preferred, so all other settings in the MCU preferences/properties, are ignored.

The advantage of such a configuration is that the versions of the binary tools are stored in the `package.json`, and exactly the same versions can be later installed if the project is cloned on a separate machine, thus being one step closer to reproducible builds.

**Comment by jonahgraham** [comment]
> In your example you show the path for the windows-build-tools, which is retrieved from the preferences as an absolute path.

This comment https://github.com/eclipse-embed-cdt/eclipse-plugins/pull/576#issuecomment-1610030583 was not in reply to your question in https://github.com/eclipse-embed-cdt/eclipse-plugins/pull/576#discussion_r1244169178 - please see https://github.com/eclipse-embed-cdt/eclipse-plugins/pull/576#discussion_r1244201803

> My question refers to the `xpacks/.bin` folder. Do you see the absolute path of this folder in the generated PATH?

Yes.

> This is a relatively recent feature, unfortunately not very well documented. :-(

AFAICT the feature currently only works if the project is checked out into the workspace folder. Also the normal paths are added to the path to.

e.g. if you have a project with `xpacks/.bin` in your project outside of the workspace the PATH gets an invalid location for `xpacks/.bin` and then get the global (or workspace) path to the toolchain. So it looks like it works, but it is not using gcc in the project's xpacks/.bin.

When `preferXpacksBin` is true should it stop adding other locations to the PATH?

**Comment by ilg-ul** [comment]
> AFAICT the feature currently only works if the project is checked out into the workspace folder.

Hmmm... not good... how can we make it work for standalone projects?

That's the point with the xpm dependencies, they are local to the project, regardless where the project is located.

**Comment by ilg-ul** [comment]
> When `preferXpacksBin` is true should it stop adding other locations to the PATH?

Not only that it should stop adding other locations to the PATH (except the windows build tools), but the `xpacks/.bin` must be guaranteed to be the very first folder in the PATH. This is how xpm/npm provide reproducible builds, and Eclipse should implement it too.

**Comment by jonahgraham** [comment]
> > AFAICT the feature currently only works if the project is checked out into the workspace folder.
> 
> Hmmm... not good... how can we make it work for standalone projects?

That is what the change I introduced fixed. It will work once you merge.







> > When `preferXpacksBin` is true should it stop adding other locations to the PATH?
> 
> Not only that it should stop adding other locations to the PATH (except the windows build tools), but the `xpacks/.bin` must be guaranteed to be the very first folder in the PATH. This is how xpm/npm provide reproducible builds, and Eclipse should implement it too.

In that case I will wrap the rest of the path code in the else block of `if (preferXpacksBin) {` - new PR on its way in a moment.

**Comment by ilg-ul** [comment]
> I will wrap the rest of the path code in the else block of if (preferXpacksBin)

be sure you keep the windows build tools, if defined, and locate them beeper than xpacks/.bin.

**Comment by jonahgraham** [comment]
I'm going to split the xpacks into a new PR.

**Comment by jonahgraham** [comment]
I have split out any of the changes related to preferXpacksBin as it was clouding the original purpose of this PR which was to fix #230.

@ilg-ul this PR is ready to go now and I will soon (but may not be today) provide a new PR for preferXpacksBin so we can discuss that.

**Review by ilg-ul — APPROVED** [review]

**Comment by ilg-ul** [comment]
I entered #580 for the changes related to preferXpacksBin.
